docs/mayavi2.man
.TH MAYAVI2 1
.SH NAME
mayavi2 \- A scientific visualization package for 2-D and 3-D data
.SH SYNOPSIS
.B mayavi2 [options]  [args]

Where arg1, arg2 etc\&. are optional file names that correspond to saved
Mayavi2 visualizations (filename\&.mv2),  Mayavi2 scripts
(filename\&.py) and any data file formats that Mayavi supports\&.

.SH "DESCRIPTION"

.PP
Mayavi2 is a tool for 2-D and 3-D scientific visualization\&. Features
of Mayavi2 include:

.PP
.IP *
Visualization of scalar and vector data in 2 and 3 dimensions
.IP *
Easy scriptability using Python 
.IP *
Easy extendability via sources, components, modules, and data filters
.IP *
Reading several file formats: VTK (legacy and XML), PLOT3D, etc\&.
.IP *
Saving of visualizations 
.IP *
Saving rendered visualization in a variety of image formats\&.


.SH "OPTIONS"
.PP
.B -h

This prints all the available command line options and exits\&.  Also
available through \fB\-\-help\fP\&.

.PP
.B -V

This prints the Mayavi version on the command line and exits\&.  Also
available through \fB\-\-version\fP\&.

.PP
.B -z filename\&.mv2

This loads a previously saved Mayavi2 visualization\&.  Also available
through \fB\-\-viz\fP filename\&.mv2 or \fB\-\-visualization\fP
filename\&.mv2\&.

.PP
.B -d
datafile\&.*

Opens any of the supported data file formats or a non-file related data
source\&.  This includes VTK file formats (*\&.vtk, *\&.xml,
*\&.vt[i,p,r,s,u], *\&.pvt[i,p,r,s,u]), VRML2 (*\&.wrl), 3D Studio
(*\&.3ds), PLOT3D (*\&.xyz), STL, BYU, RAW, PLY, PDB, SLC, FACET, OBJ,
AVSUCD (*\&.inp), GAMBIT (*\&.neu), Exodus (*\&.exii), PNG, JPEG, BMP,
PNM, DCM, DEM, MHA, MHD, MINC, XIMG, TIFF, and various others that are
supported\&.  datafile\&.* can also be a source object not associated
with a file, for example ParametricSurface or PointLoad will load the
corresponding data sources into Mayavi\&.  Also available through
\fB\-\-data\fP\&.

.PP
.B -m
module-name

A module is an object that actually visualizes the data\&.  The given
module-name is loaded in the current ModuleManager\&.  The module name
must be a valid one if not you will get an error message\&.

If a module is specified as 'package\&.sub\&.module\&.SomeModule' then the
module (SomeModule) is imported from 'package\&.sub\&.module'\&.  Standard
modules provided with mayavi2 do not need the full path
specification\&.  For example:
.PP
.RS
\f(CWmayavi2 \-d data\&.vtk \-m Outline \-m m2_user_modules\&.TestModule\fP
.RE

In this example 'Outline' is a standard module 
and 'm2_user_modules\&.TestModule' is some user defined module\&.  See
.B MAYAVI2 MODULES
section below for an exhaustive list of available modules\&.  Also
available through \fB\-\-module\fP\&.

.PP
.B -f
filter-name

A filter is an object that filters out the data in some way or the 
other\&.  The given filter-name is loaded with respect to the current
source/filter object\&.  The filter name must be a valid one if not you
will get an error message\&.

If the filter is specified as 'package\&.sub\&.filter\&.SomeFilter' then the
filter (`SomeFilter`) is imported from 'package\&.sub\&.filter'\&.  Standard
modules provided with mayavi2 do not need the full path
specification\&.  For example:
.PP
.RS
\f(CWmayavi2 \-d data\&.vtk \-f ExtractVectorNorm \-f m2_user_filters\&.TestFilter\fP
.RE

In this example 'ExtractVectorNorm' is a standard filter 
and 'm2_user_filters\&.TestFilter' is some user defined filter\&.

See
.B MAYAVI2 FILTERS
section below for an exhaustive list of available filters\&. Also
available through \fB\-\-filter\fP\&.

.PP
.B -M

Starts up a new module manager on the Mayavi pipeline\&. Also available
through \fB\-\-module\-mgr\fP\&.

.PP
.B -n

Creates a new window/scene\&. Any options passed after this will apply
to this newly created scene\&.  Also available through \fB\-\-new\-window\fP\&.

.PP
.B -t

Runs the mayavi2 test suite and exits\&.  If run as such, this runs
both the TVTK and Mayavi2 unittests\&.  If any additional arguments
are passed they are passed along to the test runner\&.  So this may be
used to run other tests as well\&.  For example:
.PP
.RS
\f(CWmayavi2 \-t enthought\&.persistence\fP
.RE

This will run just the tests inside the enthought\&.persistence
package\&.  You can also specify a directory with test files to run
with this, for example:
.PP
.RS
\f(CWmayavi2 \-t relative_path_to/integrationtests/mayavi\fP
.RE

will run the integration tests from the mayavi sources.   Also available
through \fB\-\-test\fP\&.

.PP
.B -x
script-file

This executes the given script in a namespace where we guarantee that 
the name 'mayavi' is Mayavi's script instance -- just like in the
embedded Python interpreter\&.  Also available through \fB\-\-exec\fP

.B WARNING:
Note that this uses `execfile`, so please note that
this can be dangerous if the script does something nasty!

.PP
.B -s 
python-expression

Execute the expression on the last created object\&.  For example, if
the previous object was a module, and we want to set the color of that
object one can do \-s "actor\&.property\&.color = (1,0,0)".  One should
use quotes for the expression. Also available with \fB\-\-set\fP\&.  For
example: 

.PP
.RS
 \f(CWmayavi2 \-d ParametricSurface \-m Outline \\ \fP
 \f(CW \-s "actor\&.property.color = (1,0,0)" \\ \fP
 \f(CW \-s "scene.save('test\&.png', size=(800, 800))"\fP
.RE

.B WARNING:
Note that this uses `exec`, so please note that this can be dangerous!

.SH "MAYAVI2 MODULES"

.TP
.B Axes
Draws simple axes\&.

.TP
.B ContourGridPlane
A contour grid plane module\&. This module lets one take a slice of
input grid data and view contours of the data\&.

.TP
.B CustomGridPlane
A custom grid plane with a lot more flexibility than GridPlane module\&.

.TP
.B DataSetClipper
This filter clips the dataset in a area.  The area can be
defined interactively as a box, a sphere...

.TP
.B Glyph
Displays different types of glyphs oriented and colored as per scalar 
or vector data at the input points\&.

.TP
.B GridPlane
A simple grid plane module\&.

.TP
.B HyperStreamline
A module that integrates through a tensor field to generate a 
hyperstreamline\&. The integration is along the maximum eigenvector and
the cross section of the hyperstreamline is defined by the two other 
eigenvectors\&. Thus the shape of the hyperstreamline is "tube-like",
with the cross section being elliptical. Hyperstreamlines are used to
visualize tensor fields\&.

.TP
.B ImageActor
A simple module to view image data efficiently\&.

.TP
.B ImagePlaneWidget
A simple module to view image data\&.

.TP
.B IsoSurface
A module that allows the user to make contours of input point data\&.

.TP
.B Labels
Allows a user to label the current dataset or the current actor of the 
active module\&.

.TP
.B OrientationAxes
Creates a small axes on the side that indicates the position of the
co-ordinate axes and thereby marks the orientation of the
scene\&. Requires VTK-4\&.5 and above\&.

.TP
.B Outline
A module that draws an outline for the given data\&.

.TP
.B ScalarCutPlane
Takes a cut plane of any input data set using an implicit plane and
plots the data with optional contouring and scalar warping\&.

.TP
.B SliceUnstructuredGrid
This module takes a slice of the unstructured grid data and shows the 
cells that intersect or touch the slice\&.

.TP
.B Streamline
Allows the user to draw streamlines for given vector data\&. This
supports various types of seed objects (line, sphere, plane and point
seeds)\&. It also allows the user to draw ribbons or tubes and further
supports different types of interactive modes of calculating the
streamlines\&.

.TP
.B StructuredGridOutline
Draws a grid-conforming outline for structured grids\&.

.TP
.B Surface
Draws a surface for any input dataset with optional contouring\&.

.TP
.B TensorGlyph
Displays tensor glyphs oriented and colored as per scalar or vector data at 
the input points\&.

.TP
.B Text
This module allows the user to place text on the screen\&.

.TP
.B VectorCutPlane
Takes an arbitrary slice of the input data using an implicit cut plane 
and places glyphs according to the vector field data\&. The glyphs may
be colored using either the vector magnitude or the scalar attributes\&.

.TP
.B Vectors
Displays different types of glyphs oriented and colored as per vector 
data at the input points\&. This is merely a convenience module that is
entirely based on the Glyph module\&.

.TP
.B Volume
The Volume module visualizes scalar fields using volumetric 
visualization techniques\&.

.TP
.B WarpVectorCutPlane
Takes an arbitrary slice of the input data using an implicit cut plane 
and warps it according to the vector field data\&. The scalars are
displayed on the warped surface as colors\&.


.SH "MAYAVI2 FILTERS"

.TP
.B CellDerivatives
Computes derivatives from input point scalar and vector data and 
produces cell data on the gradients\&.  Can be used to approximately
calcuate the vorticity for example\&.

.TP
.B CellToPointData
Transforms cell attribute data to point data by averaging the cell 
data from the cells at the point\&.

.TP
.B Contour
A contour filter that wraps around the Contour component to generate 
iso-surfaces on any input dataset\&.
    
.TP
.B CutPlane
This class represents a cut plane that can be used to slice through any 
dataset\&.  It also provides a 3D widget interface to position and move
the slice interactively\&.

.TP
.B ElevationFilter
Creates scalar data corresponding to the elevation of the points along a 
line\&.

.TP
.B Delaunay2D
Performs a 2D Delaunay triangulation\&.

.TP
.B Delaunay3D
Performs a 3D Delaunay triangulation\&.


docs/CHANGES.txt
Mayavi 3.3.1 (not yet released)
===============================

Enhancements
------------


Fixes
-----



Mayavi 3.3.0 (July 15, 2009)
============================

Enhancements
------------

(not collated yet, sorry)


Fixes
-----

(not collated yet, sorry)



Mayavi 3.2.0 (March 23, 2009)
=============================

A log of significant changes made to the package especially API changes.
This is only partial and only covers the developments after the 2.x
series.  The log refers to changesets in the trac format [#revno], for
example to view the changeset [18702] use the following URL:
https://svn.enthought.com/enthought/changeset/18702


17, 18 March, 2009 (PR):
    - NEW: A simple example to show how one can use TVTK's visual module
      with mlab. [23250]
    - BUG: The size trait was being overridden and was different from
      the parent causing a bug with resizing the viewer. [23243]

15 March, 2009 (GV):
    - ENH: Add a volume factory to mlab that knows how to set color, vmin and 
      vmax for the volume module [23221].

14 March, 2009 (PR):
    - API/TEST: Added a new testing entry point: 'mayavi -t' now runs
      tests in separate process, for isolation.  Added
      ``enthought.mayavi.api.test`` to allow for simple testing from the
      interpreter [23195]...[23200], [23213], [23214], [23223].
    - BUG: The volume module was directly importing the
      wx_gradient_editor leading to an import error when no wxPython is
      available. This has been tested and fixed. Thanks to Christoph
      Bohme for reporting this issue. [23191]

14 March, 2009 (GV):
    - BUG: [mlab]: fix positioning for titles [23194], and opacity for titles 
      and text [23193].
    - ENH: Add the mlab_source attribute on all objects created by mlab,
      when possible [23201], [23209].
    - ENH: Add a message to help the first-time user, using the new banner 
      feature of the IPython shell view [23208].

13 March, 2009 (PR):
    - NEW/API: Adding a powerful TCP/UDP server for scripting mayavi via
      the network. This is available in
      ``enthought.mayavi.tools.server`` and is fully documented. It uses
      twisted and currently only works with wxPython. It is completely
      insecure though since it allows a remote user to do practically
      *anything* from mayavi.

13 March, 2009 (GV)
    - API: rename mlab.orientationaxes to mlab.orientation_axes [23184]

11 March, 2009 (GV)
    - API: Expose 'traverse' in mlab.pipeline [23181]

10 March, 2009 (PR)
    - BUG: Fixed a subtle bug that affected the ImagePlaneWidget. This
      happened because the scalar_type of the output data from the
      VTKDataSource was not being set correctly. Getting the range of any
      input scalars also seems to silence warnings from VTK. This should
      hopefully fix issues with the use of the IPW with multiple scalars.
      I've added two tests for this, one is an integration test since those
      errors really show up only when the display is used. The other is a
      traditional unittest. [23166]

08 March, 2009 (GV)
    - ENH: Raises an error when the user passes to mlab an array with 
      infinite values [23150]

07 March, 2009 (PR)
    - BUG: A subtle bug with a really gross error in the GridPlane
      component, I was using the extents when I should really have been
      looking at the dimensions. The extract grid filter was also not
      flushing the data changes downstream leading to errors that are also
      fixed now. These errors would manifest when you use an ExtractGrid to
      select a VOI or a sample rate and then used a grid plane down stream
      causing very wierd and incorrect rendering of the grid plane (thanks to
      conflation of extents and dimensions). This bug was seen at NAL for a
      while and also reported by Fred with a nice CME. The CME was then
      converted to a nice unittest by Suyog and then improved. Thanks to
      them all. [23146]

28 February, 2009 (PR)
    - BUG: Fixed some issues reported by Ondrej Certik regarding the use
      Of mlab.options.offscreen, mlab.options.backend = 'test', removed cruft
      from earlier 'null' backend, fixed bug with incorrect imports,
      add_dataset set no longer adds one new null engine each time figure=False
      is passed, added test case for the options.backend test. [23088]

23 February, 2009 (PR)
    - ENH: Updating show so that it supports a stop keyword argument that
      pops up a little UI that lets the user stop the mainloop
      temporarily and continue using Python [23049]

21 February, 2009 (GV)
    - ENH: Add a richer view for the pipeline to the MayaviScene [23035]
    - ENH: Add safegards to capture wrong triangle array sizes in
      mlab.triangular_mesh_source. [23037]


enthought/mayavi/scripts/mayavi2.py
#!/usr/bin/env python
"""The Mayavi application standalone script.

This script parses the command line arguments and launches Mayavi2
suitably.  It is meant to be used by those using Mayavi2 as a
standalone application.

Mayavi2 wiki page: http://svn.enthought.com/enthought/wiki/MayaVi

Author: Prabhu Ramachandran <prabhu_r@users.sf.net>

"""
# Author: Prabhu Ramachandran <prabhu_r@users.sf.net>
# Copyright (c) 2005-2008, Enthought, Inc.
# License: BSD Style.

# Standard library imports.
import sys
import types
import getopt
import logging
from os.path import splitext, exists, join, abspath

# Local imports.
from enthought.mayavi.__version__ import __version__
from enthought.mayavi.scripts.util import get_data_dir

# Globals.
OFFSCREEN = False

# A global mayavi instance so we can close it correctly.
mayavi = None


###########################################################################
# Utility functions.
###########################################################################
def usage ():
    msg="""Usage:\n\nmayavi2 [options] ... [arg1] [arg2] ...

Where arg1, arg2 ... are optional file names that correspond to saved
Mayavi2 visualizations (file.mv2) or Mayavi2 scripts (file.py) or any
data files supported by Mayavi.  Valid options are one or more of the
following:

-d datafile.ext
--data datafile.ext

     datafile.ext can be any of the supported data file formats.  This
     includes VTK file formats (*.vtk, *.xml, *.vt[i,p,r,s,u],
     *.pvt[i,p,r,s,u]), VRML2 (*.wrl), 3D Studio (*.3ds), PLOT3D
     (*.xyz), STL, BYU, RAW, PLY,  PDB,  SLC,  FACET,  OBJ,  AVSUCD
     (*.inp),  GAMBIT (*.neu),  Exodus  (*.exii),  PNG,  JPEG,  BMP,
     PNM, DCM, DEM, MHA, MHD, MINC, XIMG, TIFF, and various others that
     are supported.  datafile.ext can also be a source object not
     associated with a file, for example ParametricSurface or PointLoad
     will load the corresponding data sources into Mayavi.

--filter filter-name
-f filter-name

     The passed filter name is loaded with respect to the current
     source/filter object.  The filter name must be a valid one if not
     you will get an error message.  

     If the filter is specified as 'package.sub.filter.SomeFilter'
     then the filter (`SomeFilter`) is imported from
     'package.sub.filter'.  Standard modules provided with mayavi2 do
     not need the full path specification.

     Example::
         mayavi2 -d data.vtk -f ExtractVectorNorm -f m2_user_filters.TestFilter

     In this example 'ExtractVectorNorm' is a standard filter and
     'm2_user_filters.TestFilter' is some user defined filter.

-m module-name
--module module-name

     The passed module name is loaded in the current ModuleManager.
     The module name must be a valid one if not you will get an error
     message.

     If a module is specified as 'package.sub.module.SomeModule' then
     the module (`SomeModule`) is imported from 'package.sub.module'.
     Standard modules provided with mayavi2 do not need the full path
     specification.

     Example::
         mayavi2 -d data.vtk -m Outline -m m2_user_modules.TestModule

     In this example 'Outline' is a standard module and
     'm2_user_modules.TestModule' is some user defined module.

-M
--module-mgr

     Starts up a new module manager on the Mayavi pipeline.

-n
--new-scene

     Creates a new TVTK scene.  Any options passed after this will
     apply to this newly created scene.

-o
--offscreen

     Run Mayavi in offscreen mode without any graphical user interface.
     This is most useful for scripts that need to render images
     offscreen.

-x script-file
--exec script-file

     This executes the given script in a namespace where we guarantee
     that the name 'mayavi' is Mayavi's script instance -- just like
     in the embedded Python interpreter.

     **WARNING**: Note that this uses `execfile`, so please note that
                  this can be dangerous if the script does something
                  nasty!

-s python-expression
--set python-expression

     Execute the expression on the last created object.  For example,
     lets say the previous object was a module.  If you want to set the
     color of that object and save the scene you may do::

      $ mayavi2 [...] -m Outline -s"actor.property.color = (1,0,0)" \
        -s "scene.save('test.png', size=(800, 800))"

     You should use quotes for the expression.

     **WARNING**: Note that this uses `exec`, so please note that
                  this can be dangerous!

-z saved-visualization-file
--viz saved-visualization-file
--visualization saved-visualization-file

     Loads a previously saved Mayavi2 visualization file passed as the
     argument.

-t
--test
    
     Runs the mayavi2 test suite and exits.  If run as such, this runs
     both the TVTK and Mayavi2 unittests.  If any additional arguments
     are passed they are passed along to the test runner.  So this may
     be used to run other tests as well.  For example::

        mayavi2 -t enthought.persistence

     This will run just the tests inside the enthought.persistence
     package.  You can also specify a directory with test files to run
     with this, for example::

        mayavi2 -t relative_path_to/integrationtests/mayavi

     will run the integration tests from the mayavi sources.

-v
--verbose

     Prints verbose logs on the console.

-V 
--version

     Prints the Mayavi version.

-h
--help

     Prints this help message.

Examples::

     mayavi2 test.mv2 (or mayavi2 -z test.mv2)

     mayavi2 test.py (or mayavi2 -x test.py)

     mayavi2 test.mv2 other.mv2 (or mayavi2 -z test.mv2 -z other.mv2)

     mayavi2 -d test.vtk -m Axes -m GridPlane \\
     -f Threshold -m IsoSurface \\
     -n -d heart.vtk -m Outline -m ContourGridPlane 
"""
    return msg


def parse_cmd_line(arguments):
    """Returns the options and arguments parsed via getopts.

    Due to the special needs (order of args is important and repeated
    options will be supplied) of this command line parser we must use
    getopts and not optparse.
    
    Input Arguments:
      arguments -- This can be either a list of arguments as in      
                   sys.argv[1:] or a string that is similar to the one
                   passed on the command line.  If it is a string the
                   string is split to create a list of arguments.
    """
    if type(arguments) in types.StringTypes:
        arguments = arguments.split()
        
    options = "d:m:f:z:x:s:nMvo"
    
    long_opts = ['data=',
                 'module=', 'filter=',
                 'visualization=', 'viz=', 
                 'exec=',
                 'set=',
                 'verbose',
                 'module-mgr', 'new-scene', 'offscreen']

    try:
        opts, args = getopt.getopt (arguments, options, long_opts)
    except getopt.error, msg:
        print msg
        print usage ()
        print '-'*70
        print msg
        sys.exit (1)
        
    return opts, args


def _get_non_file_sources():
    """Returns a dict indexed on the name of non-file related sources
    ids with the value being the corresponding metadata object.  
    """
    from enthought.mayavi.core.registry import registry
    data = {}
    for src in registry.sources:
        if len(src.extensions) == 0:
            name = src.id[:-6]
            data[name] = src
    return data


def process_cmd_line(app, opts, args):
    """ Processes the passed command line arguments.

    Input Arguments:
      app -- A Mayavi application instance.

      opts -- The list of options returned by getopt.

      args -- The remaining arguments returned by getopt.
    """    

    from enthought.mayavi.core.common import error, exception
    from enthought.tvtk.common import camel2enthought

    sources = _get_non_file_sources()
    script = app.script
    last_obj = None

    # Start a new scene by default if there is none currently and none
    # was specified at the start of the command line arguments.
    if script.engine.current_scene is None:
        new_scene = False
        if len(opts) == 0:
            if len(args) == 0:
                new_scene = True
        elif (opts[0][0] not in ('-n', '--new-scene', '-z',
                                 '--visualization', '--viz',
                                 '-x', '--exec')):
            new_scene = True
        if new_scene:            
            last_obj = script.new_scene()
        
    for o, a in opts:
        if o in ('-d', '--data'):
            base, ext = splitext(a)
            if exists(a):
                last_obj = script.open(a)
            elif a in sources:
                md = sources[a]
                src = md.get_callable()()
                script.add_source(src)
                last_obj = src
            else:
                error("File/Source %s does not exist!"%a)
                return

        if o in ('-m', '--module'):
            if '.' in a:
                idx = a.rfind('.')
                modname = a[:idx]
                classname = a[idx+1:]
            else:
                modname = 'enthought.mayavi.modules.%s'%camel2enthought(a)
                classname = a
            try:
                mod = __import__(modname, globals(), locals(), [classname])
            except ImportError, msg:
                exception(str(msg))
                return
            else:
                m = getattr(mod, classname)()
                if classname == 'Labels':
                    m.object = script.engine.current_object
                script.add_module(m)
                last_obj = m

        if o in ('-f', '--filter'):
            if '.' in a:
                idx = a.rfind('.')
                modname = a[:idx]
                classname = a[idx+1:]
            else:
                if a[:12] == 'UserDefined:':
                    modname = 'enthought.mayavi.filters.user_defined'
                    classname = 'UserDefined'
                    # Create the wrapped filter.
                    fname = a[12:]
                    from enthought.tvtk.api import tvtk
                    try:
                        extra = getattr(tvtk, fname)()
                    except (AttributeError, TypeError):
                        # Don't worry about errors.
                        extra = None
                else:
                    modname = 'enthought.mayavi.filters.%s'%camel2enthought(a)
                    classname = a
                    extra = None
            try:
                mod = __import__(modname, globals(), locals(), [classname])
            except ImportError, msg:
                exception(str(msg))
                return
            else:
                klass = getattr(mod, classname)
                if classname != 'UserDefined':
                    f = klass()
                else:
                    if extra is not None:
                        f = klass(filter=extra)
                    else:
                        f = klass()
                    f.setup_filter()
                script.add_filter(f)
                last_obj = f

        if o in ('-M', '--module-mgr'):
            from enthought.mayavi.core.module_manager \
                 import ModuleManager
            mm = ModuleManager()            
            script.add_filter(mm)
            last_obj = mm

        if o in ('-n', '--new-scene'):
            script.new_scene()
            e = script.engine
            s = e.scenes[-1]
            e.set(current_scene=s, current_object=s)
            last_obj = s

        if o in ('-x', '--exec' ):
            err = run_script(script, a)
            if err: # stop processing options.
                return

        if o in ('-s', '--set'):
            try:
                stmt = 'last_obj.' + a
                exec stmt in locals(), globals()
            except Exception, msg:
                exception(str(msg))

        if o in ('-z', '--visualization', '--viz'):
            script.load_visualization(a)

    # for remaining arguments simply load saved visualizations.
    for arg in args:
        base, ext = splitext (arg)
        if ext == '.mv2':
            script.load_visualization(arg)
        elif ext == '.py':
            err = run_script(script, arg)
            if err: # stop processing arguments.
                return
        else:
            script.open(arg)


def run_script(mayavi, script_name):
    """Execfiles a given script.  The name `mayavi` is bound to the
    mayavi script instance just like in the embedded interpreter.
    `script_name` is the name of the script to execute.

    Note that this function uses `execfile`. You should be careful
    when using this.

    It returns `False` if everything was OK and `True` if not.
    """
    from enthought.mayavi.core.common import exception

    g = sys.modules['__main__'].__dict__
    if 'mayavi' not in g:
        g['mayavi'] = mayavi
        g['engine'] = mayavi.engine
    error = False
    # Do execfile
    try:
        # If we don't pass globals twice we get NameErrors and nope,
        # using exec open(script_name).read() does not fix it.
        execfile(script_name, g, g)
    except Exception, msg:
        exception(str(msg))
        error = True

    return error

# This runs the runtests script and sends any args to it.
if ('-t' in sys.argv[1:]) or ('--test' in sys.argv[1:]):
    from enthought.mayavi.tests import runtests
    for arg in ('-t', '--test'):
        if arg in sys.argv[1:]:
            sys.argv.remove(arg)
    runtests.main()

# If the user just wants help messages.  Print them before importing
# any of the big modules.
if ('-h' in sys.argv[1:]) or ('--help' in sys.argv[1:]):
    print usage()
    sys.exit(0)

if ('-V' in sys.argv[1:]) or ('--version' in sys.argv[1:]):
    print 'Mayavi %s'%__version__
    sys.exit(0)

for opt, arg in parse_cmd_line(sys.argv[1:])[0]:
    if opt in ('-o', '--offscreen'):
        OFFSCREEN = True
        break
# Create opt and arg to be able to delete them even if the previous loop
# is empty.
opt, arg = None, None
del opt, arg


# The vtk module is imported by the engine, so mayavi is never going to 
# start without it. Let us capture the error early, and give a meaningful
# error message
try:
    import vtk
except ImportError, m:
    msg = '%s\n%s\nDo you have vtk installed properly?\n' \
        'VTK (and build instructions) can be obtained from http://www.vtk.org\n' \
        % (m, '_'*80)
    raise ImportError(msg)

try:
    # Try forcing the use of wx 2.8
    from enthought.etsconfig.api import ETSConfig
    if ETSConfig.toolkit in ('wx', ''):
        import wxversion
        if wxversion.checkInstalled('2.8'):
            try:
                wxversion.select('2.8')
            except wxversion.VersionError:
                """ Too late, wxPython has already been imported. """
except ImportError:
    """ wxversion not installed """


# Importing here to avoid time-consuming import when user only wanted
# version/help information.
try:
    from enthought.mayavi.plugins.app import Mayavi, setup_logger
except ImportError, m:
    msg = '''%s
%s
Could not load envisage. You might have a missing dependency.
Do you have the EnvisageCore and EnvisagePlugins installed?

If you installed Mayavi with easy_install, try 'easy_install <pkg_name>'.
'easy_install Mayavi[app]' will also work.

If you performed a source checkout and installed via 'python setup.py develop',
be sure to run the same command in the EnvisageCore and EnvisagePlugins folders.

If these packages appear to be installed, check that your numpy and
configobj are installed and working. If you need numpy, 'easy_install numpy' 
will install numpy. Similarly, 'easy_install configobj' will install 
configobj.
        ''' % (m, '_'*80)
    raise ImportError(msg)



##########################################################################
# `MayaviApp` class
##########################################################################
class MayaviApp(Mayavi):
    def parse_command_line(self, argv):
        """Parse command line options."""
        # Parse and store command line options to process once app has
        # started in `run`.
        options, args = parse_cmd_line(argv)
        self.cmd_line_opts = (options, args)
        # If the verbose option is set, change the log mode.
        for opts, args in options:
            if opts in ('-v', '--verbose'):
                self.log_mode = logging.DEBUG
                break

    def run(self):
        """Process the command line options and setup mayavi as per
        the users needs.
        """
        options, args = self.cmd_line_opts
        # Process the options.
        process_cmd_line(self, options, args)


################################################################################
# `MayaviOffscreen` class.
################################################################################ 
class MayaviOffscreen(MayaviApp):
    """
    The mayavi application used for offscreen rendering.
    """

    def _script_default(self):
        from enthought.mayavi.plugins.script import Script
        from enthought.mayavi.core.off_screen_engine import OffScreenEngine
        engine = OffScreenEngine()
        engine.start()
        s = Script(engine=engine)
        return s

    def setup_logger(self):
        from enthought.etsconfig.api import ETSConfig
        path = join(ETSConfig.application_data, 
                    'enthought.mayavi_e3', 'mayavi.log')
        path = abspath(path)
        logger = logging.getLogger()
        setup_logger(logger, path, mode=self.log_mode)        

    def main(self, argv=None):
        if argv is None:
            argv = []

        self.parse_command_line(argv)
        self.setup_logger()
        self.run()


##########################################################################
# Helper functions
##########################################################################
def get_mayavi_script_instance():
    """Return the mayavi Script instance from the first available set of
    envisage engines registered in the registry.
    """
    from enthought.mayavi.core.registry import registry
    from enthought.mayavi.plugins.envisage_engine import EnvisageEngine
    from enthought.mayavi.plugins.script import Script
    for name, engine in registry.engines.iteritems():
        if isinstance(engine, EnvisageEngine):
            return engine.window.get_service(Script)
    return

def contains_mayavi(namespace):
    """Returns if the given namespace contains a 'mayavi' name bound to
    a mayavi script instance.
    """
    from enthought.mayavi.plugins.script import Script
    if 'mayavi' in namespace:
        if isinstance(namespace.get('mayavi'), Script):
            return True
    return False

def standalone(func):
    """A decorator to run a function from within mayavi.  This lets
    users write a normal Python function and have that run from within
    mayavi.  It implicitly assumes that the name 'mayavi' refers the the
    Script instance and will overwrite it if not.
    """
    def wrapper(*args, **kw):
        script = get_mayavi_script_instance()
        if script is None and mayavi is not None:
            script = mayavi.script
        if script is None:
            def caller(script):
                """Callback that runs the function inside the mayavi
                app."""
                # Bind the 'mayavi' name to the script instance
                func.func_globals['mayavi'] = script
                # Run the function in the event loop.
                g = script.window.application.gui
                g.invoke_later(func, *args, **kw)

            # Start up mayavi and invoke caller when the script instance
            # is available.
            m = Mayavi()
            m.on_trait_change(caller, 'script')
            # Run the mayavi app.
            m.main()
        else:
            ns = func.func_globals
            if not contains_mayavi(ns):
                # Bind the 'mayavi' name to the script instance
                ns['mayavi'] = script
            # Now run the function.
            func(*args, **kw)

    return wrapper


def main():
    """This starts up the mayavi2 application.
    """
    global mayavi

    # Make sure '.' is in sys.path
    if '' not in sys.path:
        sys.path.insert(0, '')
    # Start the app.
    if OFFSCREEN:
        mayavi = MayaviOffscreen()
    else:
        from enthought.etsconfig.api import ETSConfig
        # Check that we have a traits backend installed
        from enthought.traits.ui.toolkit import toolkit
        toolkit() # This forces the selection of a toolkit.
        if ETSConfig.toolkit in ('null', ''):
            raise ImportError, '''Could not import backend for traits
________________________________________________________________________________
Make sure that you have either the TraitsBackendWx or the TraitsBackendQt
projects installed. If you installed Mayavi with easy_install, try easy_install 
<pkg_name>. easy_install Mayavi[app] will also work.

If you performed a source checkout, be sure to run 'python setup.py install'
in Traits, TraitsGUI, and the Traits backend of your choice.

Also make sure that either wxPython or PyQT is installed.
wxPython: http://www.wxpython.org/
PyQT: http://www.riverbankcomputing.co.uk/software/pyqt/intro
'''
            
        mayavi = MayaviApp()
    mayavi.main(sys.argv[1:])

def close():
    """ This closes the mayavi2 application.
    """
    if mayavi is not None and not OFFSCREEN:
        mayavi.window.close()

if __name__ == '__main__':
    main()
